providers/providers_test.go

```diff
@@ -26,22 +26,59 @@ func TestProviderManager(t *testing.T) {
 	defer cancel()
 
 	mid := peer.ID("testing")
-	p := NewProviderManager(ctx, mid, dssync.MutexWrap(ds.NewMapDatastore()))
-	a := u.Hash([]byte("test"))
-	p.AddProvider(ctx, a, peer.ID("testingprovider"))
+	pm := NewProviderManager(ctx, mid, dssync.MutexWrap(ds.NewMapDatastore()))
+	key := u.Hash([]byte("test"))
+	pm.AddProvider(ctx, key, peer.ID("first provider"))
 
 	// Not cached
-	resp := p.GetProviders(ctx, a)
+	resp := pm.GetProviders(ctx, key)
 	if len(resp) != 1 {
 		t.Fatal("Could not retrieve provider.")
 	}
 
 	// Cached
-	resp = p.GetProviders(ctx, a)
+	resp = pm.GetProviders(ctx, key)
 	if len(resp) != 1 {
 		t.Fatal("Could not retrieve provider.")
 	}
-	p.proc.Close()
+
+	// Make sure that cache gets updated
+	pm.AddProvider(ctx, key, peer.ID("second provider"))
+	pm.AddProvider(ctx, key, peer.ID("third provider"))
+	pm.AddProvider(ctx, key, peer.ID("fourth provider"))
+
+	resp = pm.GetProviders(ctx, key)
+	if len(resp) != 4 {
+		t.Fatalf("Should have 4 providers. Got %d\n", len(resp))
```

Comment:
This test passes on master as well. What was the observed behavior?

Reply:
I'm not being able to reproduce the error anymore :disappointed:  Already started the PR from scratch and did all the steps that led me to identify it in the first place and despite having reproduced it several times in the morning, now it always passes flawlessly. Apologies for making you waste your time

Reply:
Np. I wonder if it was some kind of interaction with garbage provider record collection?